File: Assignment 5/Python/04. insertion & deletion at a position in a singly circular ll .py

class node:
    def __init__(self,val):
        self.val = val
        self.next = None

def insertion(head, val, pos):
    nn = node(val)
    ptr = head;
    if(pos==1):
        while ptr.next!=head:
            ptr = ptr.next
        ptr.next = nn
        nn.next = head
        return nn
    
    for i in range(2,pos):
        ptr = ptr.next

    nn.next = ptr.next
    ptr.next = nn
    return head


def deletion(head,pos):
    ptr = head
    if(pos==1):
        while(ptr.next!=head):
            ptr = ptr.next;
        
        ptr.next = head.next
        head.next = None
        return ptr.next
    
    for i in range(2,pos):
        ptr = ptr.next

    temp = ptr.next
    ptr.next = ptr.next.next
    temp.next = None
    return head


def display(head):
    ptr = head
    while True:
        print(ptr.val,end=" ")
        ptr = ptr.next
        if(ptr==head):
            break
    print("\n")

head = node(1)
snd = node(2)
trd = node(3)
four = node(4)

head.next = snd
snd.next = trd
trd.next = four
four.next = head

print("before: ")
display(head)

print("After: ")
head = insertion(head,99,4)
display(head)

deletion(head,4)
print("after deletion: ")
display(head)



# OUTPUT:
# before:
# 1 2 3 4
# After:
# 1 2 3 99 4
# after deletion:
# 1 2 3 4

File: Assignment 5/C/04. insertion & deletion at a position in a singly circular ll.c
#include <stdio.h>
#include <stdlib.h>

typedef struct node{
    int val;
    struct node* next;
}node;

void traversal(node* head){
    node* ptr = head;
    do{
        printf("--> %d ",ptr->val);
        ptr = ptr->next;
    }while(ptr!=head);
    printf("--\n");
}

node* insertion(node* head,int val,int pos){
    node* nn = (node*)malloc(sizeof(node));
    nn->val = val;
    node* ptr = head;
    if(pos==1){
        while(ptr->next!=head){
            ptr = ptr->next;
        }
        ptr->next = nn;
        nn->next = head;
        return nn;
    }
    for(int i=2; i<pos; i++){
        ptr = ptr->next;
    }
    nn->next = ptr->next;
    ptr->next = nn;
    return head;
}

node* deletion(node* head,int pos){
    node* ptr = head;
    if(pos==1){
        while(ptr->next!=head){
            ptr = ptr->next;
        }
        ptr->next = head->next;
        head->next = NULL;
        free(head);
        return ptr->next;
    }
    for(int i=2; i<pos; i++){
        ptr = ptr->next;
    }
    node* temp = ptr->next;
    ptr->next = ptr->next->next;
    temp->next = NULL;
    free(temp);
    return head;
}

int main(){
    node* head = (node*)malloc(sizeof(node));
    node* snd = (node*)malloc(sizeof(node));
    node* trd = (node*)malloc(sizeof(node));
    node* four = (node*)malloc(sizeof(node));

    head->val = 1;
    snd->val = 2;
    trd->val = 3;
    four->val = 4;

    head->next = snd;
    snd->next = trd;
    trd->next = four;
    four->next = head;

    head = insertion(head,99,1);
    traversal(head);

    head = deletion(head,3);
    traversal(head);

    return 0;
}
